### PR TITLE
added "file nesting" support, based on file naming

### PR DESCRIPTION
**Pull request by playerx**
Implemented "File Nesting" feature, which was requested by community #6328.

Also created new 'explorer.fileNesting' setting and user can enable or disable it.

screen shot 2016-10-14 at 16 51 26

### CODE REVIEW

**Comment by bpasero** [comment]
@playerx how do you decide if a file should be grouped below another file? just by checking for other files with the same file name and extension with a suffix? I am not sure that the feature request can be solved with such a simple assumption. The solution explorer in VS shows a logical view of the project based on many rules.

**Comment by playerx** [comment]
hello @bpasero, yes rule is based on file name and extension suffix.

this feature is implemented in Visual Studio Professional by default. And it's the important feature to use Visual Studio Code with Angular2.

The solution explorer in VS can show a logical view of the project based on many rules, but this will be additional level, on top of that, if user enables this setting.

P.S. Maybe, 'fileNestingByNaming' could be better name for this setting, what do you think?

**Comment by bpasero** [comment]
@playerx I am asking because a typical use case is to show derived resources below the file name as nested resource. These derived resources can live anywhere on the workspace in a different folder (e.g. out folder). I think if we ever decide to show this structure in the explorer, we need to take a look at other use cases and find a solution that helps for all of them.

My personal feeling is that the file explorer should always show the truth as it is on disk and a logical view is a separate explorer that understands the project nature of a workspace.

**Comment by playerx** [comment]
I agree with your personal feelings, and this feature makes explorer better viewer, it doesn't change files structure, it still shows you the truth, with more elegant way IMHO.

This feature can be disabled by default and developers will enable it, if they need of course.  (#6328)

**Comment by playerx** [comment]
P.S. And showing derived resources that can live anywhere on the workspace is absolutely different feature.

**Comment by bpasero** [comment]
@playerx it may be a different feature for your scenario but I find it not very practical to introduce a setting for each kind of logical view the explorer could have. I am trying to move towards a solution that makes most users happy, and I would like to avoid introducing a solution for a very specific structure.

Nevertheless this is a change that needs UX talking and PM steering to decide if we want to push into this direction.

Meanwhile thanks for this PR it can help start the discussion 👍

**Comment by playerx** [comment]
Okey, thanks.

**Comment by playerx** [comment]
@bpasero we can do following:

1. Change file nesting logic and make it based on different regular expressions. There will be an array of regular expressions and it will be applied to files structure.
2. This array will be accessible through extensions API
3. Everyone will be able to write own extensions to apply there own reg expressions

What do you think about that? Will it be Okey?

**Comment by playerx** [comment]
Example of regular expressions:
**Scenario1:**
```
component.ts
- component.ts.html
- component.ts.scss
```
regexp:
`{{filename}}\.{{ext}}\.(.+)`

**Scenario2:**
```
webpack.config.js
- webpack.config.common.js
- webpack.config.dev.js
- webpack.config.prod.js
```
regexp:
`{{filename}}\.(.+)\.{{ext}}`

**Comment by playerx** [comment]
@bpasero can you give me some feedback?

**Comment by bpasero** [comment]
We are in end game currently, I will only be able to come back to this next month 👍

**Comment by uxsoft** [comment]
Would it be possible to extend this to support nesting for typescript products (example: app.js, app.d.ts and app.js.map under app.ts)? This scenario should also be representable by RegEx so if the RegEx would be available in settings or even better if user would be able to specify several regexes for nesting:-)

**Comment by joh1234** [comment]
Update README
https://goo.gl/JBljdR

**Comment by wrod7** [comment]
will this be in 1.9?

**Comment by tommck** [comment]
any updates on this?

**Comment by ciel** [comment]
So does this work or not? I'm confused, and can't seem to make this work. Is there any update on the possibility to have more nesting schemes?

**Comment by ciel** [comment]
The biggest reason I want this is for angular, where I have many similar named files together;

```
/app
.../home.ts
.../home.component.ts
.../home.template.html
.../home.styles.scss
.../home.component.spec.ts
.../nav.ts
.../nav.component.ts
.../nav.component.spec.ts
.../nav.template.html
.../nav.styles.scss
```

Man, if it could nest those, it'd be just great.

I also do Command/Mediator pattern, and so I end up with files that look like ...

```
/src/
.../Identity/
....../CreateUserRequest.cs
....../CreateUserRequestHandler.cs
....../ConfirmUserExistsRequest.cs
....../ConfirmUserExistsRequestHandler.cs
....../ConfirmUserValidationRequest.cs
....../ConfirmUserValidationRequestHandler.cs
```

It gets pretty verbose and man, this feature would just make my day.

**Comment by firelizzard18** [comment]
IMO, the most elegant solution would be to leave it up to the project. As in `.vscode/settings.json` or maybe `project.json`. Use something like the file exclusion syntax. This would allow for per-project or per-folder customization. Framework/tool maintainers, or VSCode, or a third party could distribute pattern sets for given workflows.

This way, VSCode assumes nothing and causes issues for no one. It could detect possible candidates for nesting, and pop up an alert asking, "It looks like you have files that should be nested. Do you want to apply one of these default nesting profiles?"

If developers want logical nesting, then can have that. If they don't want it, they just don't add the settings.

@ciel, for capturing `my-component.A.B` under `my-component.ts`:
```
"files.nest": {
    "**/*.*.*": {"when": "$(basename).ts"}
}
```

@playerx, for capturing `my-component.ts.A` under `my-component.ts`:
```
"files.nest": {
    "**/*.ts.*": {"when": "$(basename).ts"}
}
```

Or for generated files (using the Aurelia navigation skeleton where files are generated into `dist/`), for TypeScript => ES5:
```
"files.nest": {
    "dist/src/**/*.js": {"when": "src/$(dir)/$(basename).ts"}
}
```

ESNext => ES5:
```
"files.nest": {
    "dist/src/**/*.js": {"when": "src/$(dir)/$(basename).js"}
}
```

Less => CSS:
```
"files.nest": {
    "dist/src/**/*.css": {"when": "src/$(dir)/$(basename).less"}
}
```

Mappings:
```
"files.nest": {
    "dist/src/**/*.map": {"when": "src/$(dir)/$(basename)"}
}
```

With some well-defined syntax like this, I could add whatever nesting scheme I wanted, as long as it's not excessively complicated. The last example is based on gulp's handling of paths (`$(path)` = the path starting with the first glob).

**Comment by itanex** [comment]
Has this been abandoned?

**Comment by tommck** [comment]
hope not. It would be a great feature

**Comment by jayoungers** [comment]
I've been following #6328 for a long time now and just assumed it was a change that would require large alterations at the base of the product.  

I didn't realize this was here until @firelizzard18 referenced it in that issue:  it blows my mind this has been out here for a year, and can be done at a basic level in 100 lines of code.

I can understand the thought that the current on/off implementation is too opinionated, but we really couldn't come to an agreement on some way of getting it implemented using the settings file?

**Comment by firelizzard18** [comment]
TL;DR: as of yet, no one has cared enough to make it happen.

@jayoungers Here's my take. @playerx made three commits within a day of each other, nearly a year ago, and nothing since. So (s)he doesn't seem to be interested any more. The VSCode team seems mildly interested, but it's low on their priority list and no consensus of how to do it has emerged.

So at this point, it's up to you; as in, it's up to individual contributors deciding to do it themselves. I made my proposal. I think it's a good solution. But unless this picks up again, nothing will happen without someone deciding to do it themselves.

I may fork this PR or make my own and attempt to implement my proposal.

**Comment by firelizzard18** [comment]
@itanex @tommck @jayoungers I created a more up to date PR: #32061.

I am looking for help as my handling of some things (namely tree model) is buggy.

**Review by tmarkovski — APPROVED** [review]

**Comment by tommck** [comment]
So, these were approved, but the code needs to be updated?

**Comment by tmarkovski** [comment]
@playerx Would you be able to update the branch against latest master?

**Comment by playerx** [comment]
Hi @tmarkovski, 

Sure! I will update the branch, but it was long ago (in 2016), and there will be lots of changes I think, it may take some time. I'll do my best.

**Comment by playerx** [comment]
Happy new year everyone! I've updated PR, added basic support for `File Nesting` feature.

@tmarkovski Please check out when you will have a chance.

Idea is simple: Adding new Virtual Layer on top of existing file tree and mark every element as: `Virtual Directory | Virtual Directory Member | None`

It's a basic support because I do filename matching by hardcoded regex patterns, I think this configuration will be one more configuration in the settings, next to `enableFileNesting` option

Please share the feedback if you like it or if I did something really bad, Im ready to fix it.

If you will have any suggestions about improving UX it will be great, for now when you click the file (which has nested files) it will be opened and expanded also.

**Comment by tmarkovski** [comment]
Thank you @playerx for the fast update.

@Tyriar This PR has been been updated and is for a highly requested feature by the community tracked at #6328 . Would you be able to get this through the review process or assign to a member of your team?
Thank you!

**Comment by playerx** [comment]
@chrisdias @Tyriar 

Status Update:
* Tested the feature and fixed couple of issues, now it works well in different use cases
* Refactored codebase made naming of the variables consistent
* Added new setting `fileNestingPatterns` with default options

Waiting for your review, please let me know if there are more changes necessary, thanks

**Comment by isidorn** [comment]
Thank you for you PR.
I see how File Nesting  is an important feature, and we definetely want to add this in the future. However the current approach which you did is not the way I think this should be done. There should just be one setting to turn this on, and that is it.
This might need work from the tree. And might need an different approach alltogetrher -> anyways this requires planning, and I might look more into this in the near future. The comment from @bpasero [here](https://github.com/microsoft/vscode/pull/13754#issuecomment-253798839) still holds true

Sorry if you put too much work into this PR before I commented, this was only approved and driven by the community.

**Comment by playerx** [comment]
@isidorn thanks for the response, Im ready to participate in the planning process if you will find it useful

I understand what you are saying. I've covered all use cases that is requested by the community in this PR, but we need more generic solution.

I think we are talking about two different feature:
1. File Nesting - simple and useful feature that's requested by the community
2. Feature you've described - which is generic, covers many use cases, including `Feature 1`, but isn't requested by the community

We have a situation that `Feature 2` is a blocker for the `Feature 1`, since 2016.

Maybe it's worth to have `Feature 1` merged, which is requested by many of us, next start planning `Feature 2` and once it will be ready, `Feature 1` will be automatically replaced. Maybe this planning process will take next 4 years and during this period we will be able to use simple `File Nesting` feature.

I've shared the same idea to @bpasero 4 years ago, but after that we didn't hear anything from VSCode team. Hope you will be more pragmatic this time, rather than idealistic.

**Comment by tommck** [comment]
Yeah, honestly, we've waited 4 YEARS and now we're told the approach is wrong?
Can we get this feature and then refactor the implementation?

**Comment by tommck** [comment]
Is there a way this could be turned into an _extension_ so that we don't have to wait 4 MORE years for this feature??

**Comment by tommck** [comment]
@isidorn ? Is it possible to do this type of thing in an extension?

**Comment by isidorn** [comment]
No. Unless the extension contributes the whole explorer view.

**Comment by tommck** [comment]
> No. Unless the extension contributes the whole explorer view.

So... we're effectively stuck with no options?

**Comment by yringler** [comment]
> > No. Unless the extension contributes the whole explorer view.
> 
> So... we're effectively stuck with no options?

Fork?...
It's a highly sought after and high profile feature. Maybe there's a chance that a fork could catch on.

**Comment by RussKie** [comment]
@bpasero perhaps it is worth chatting to folks in the VS land who have a similar feature implemented to have a unified and coherent approach.

**Comment by RussKie** [comment]
@bpasero perhaps it is worth chatting to folks in the VS land who have a similar feature implemented to have a unified and coherent approach.
https://docs.microsoft.com/en-us/visualstudio/ide/file-nesting-solution-explorer?view=vs-2019

**Comment by HosseinSalmanian** [comment]
@isidorn 
It`s so annoying to see bunch of logicaly related file in explorer view.
after almost 5 years it must be the time to change the priority of such a feature according to community requests.

**Comment by niceguyalberto** [comment]
To me the lack of file nesting is quite annoying and it's a pity that after nearly 5 years nothing has happened yet...
(unfortunately I do not have the time/skills to contribute with such a modification)

**Comment by markadrake** [comment]
I'll copy and paste something I recently shared in https://github.com/microsoft/vscode/issues/6328#issuecomment-877634046:

> I recently started a .NET 5 project and learned that the old way they did file nesting no longer worked. Seems now they use a configuration file .filenesting.json that has a proper schema and everything.
>
> I know this issue (and solution) has been seriously delayed but for compatibility with big brother Visual Studio and big sister .NET Core I'd suggest that a future solution follow the same pattern...

It seems to me a pull request to enable file nesting that follows this exact pattern **should** be easier for the maintainers to accept. This pattern is robust, documented, and in-use across .NET Core / .NET development.

If we can't get traction on this by taking action (and I commend you for making this PR) and can't make traction by offering monetary rewards (trust me, I've tried) - perhaps someone smarter than myself could look at the wizardy of the Synthwave package. Perhaps this is a way we can deliver an extension that modifies the explorer view without fully replacing it?

**Comment by isidorn** [comment]
No plans to merge, thus closing.
